add parser of jira issues

ecklbarb opened the pull request.

ecklbarb committed:
add parser of jira issues

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+    return issues
+
+
+def print_to_disk(issues, results_folder):
```
Do we still need this function? We have the function `print_to_disk_extr`...

ecklbarb replied:
The difference between this two functions is that print_to_disk includes the original issue and all comments, in contrast print_to_disk_extr incluedes only the comments of a issue.

clhunsen replied:
Ah, thanks. I guess, I have seen that later after writing this comment, but forgot to remove it then...

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+    print_to_disk(issues, __resdir)
+    # 5) export for Gephi
+    print_to_disk_gephi(issues, __resdir)
+    # 6) export for github issue extraction
```
It's Jira here and not GitHub.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+    # run processing of issue data:
+	#yxz = load_csv(__psrcdir)
+    #persons = parse_csv(yxz)
```
Please remove the commented lines here. *Or* put a comment in front of both chunks here, stating "this is only needed for ..." or "this is not needed right now, because ...".

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+from os import listdir
+from os.path import isfile, join
+from xml.dom.minidom import parse, parseString
```
Please do not import functions like `parse` with their bare name. That is heavily ambiguous, as we may need to parse several things with different functions sometime. 

Additionally, I would not import anything from `os` by their names, always use the fully qualified name (e.g., `os.path.join`).

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+    # get source and results folders
+    folder = args.resdir
+    __srcdir = os.path.abspath(os.path.join(folder, __conf['repo'] + "_proximity", "conway", "issues_xml"))
```
Why do you a separate variable (`folder`) for using it just once? Stick to `args.resdir` instead.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+    # 6) export for github issue extraction
+    print_to_disk_extr(issues, __resdir)
+
+    log.info("Issue processing complete!")
```
Please incorporate "Jira" as an string in such log messages. If you have time, please improve the log messages in the other file by adding "GitHub".

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+    :param source_folder: the folder where to find .xml-files
+    :return: the loaded issue data
+    """
```
Empty line after this line. Additonally, add a comment on the list comprehension after that.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+    return issue_data
+
+
+def load_csv(source_folder):
```
This function is not called anywhere. If you want to keep it, comment it or move it below and add a 'unneeded right now' comment.

ecklbarb replied:
Now this function is used to load and parse the csv-file which includes the names, user names and e-mail-adresses.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+    if not os.path.exists(srcfile):
+        log.info("Person file '{}' does not exist! Exiting early...".format(srcfile))
+        sys.exit(-1)
+
```
Remove empty line.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+     # check if file exists and exit early if not
+    if not os.path.exists(srcfile):
+        log.info("Person file '{}' does not exist! Exiting early...".format(srcfile))
```
`log.error()` or something similar should be available.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+
+    person_data = list()
+     #with open(srcfile, 'r') as person_file:
```
Obsolete line.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+    return person_data
+
+
+def parse_csv(person_date):
```
Same thing as for `load_csv`: Move, comment out, or delete.
Additionally, the function is dysfunctional as the return value `persons` is not defined.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+def parse_csv(person_date):
+    for row in person_date:
+        print(row['AuthorID'], row['userEmail'])
```
No `print` calls. We have a logger for that if you really need to log something.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+        :param issue_data: xml-data
+        :return: the loaded issue data
+        """
```
Watch the indentation here. Python might dislike that sometimes.

ecklbarb replied:
I'm sorry. I don't understand what do you mean.

clhunsen replied:
The indentation of the comment is messed up, i.e., the triple-quotes `"""` are not well-aligned, and neither is the content in relation to that.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+            assignee = issue_x.getElementsByTagName('reporter')[0]
+            user = dict()
+            user["username"] = assignee.attributes['username'].value
+            user["name"] = assignee.attributes['username'].value	##assignee.firstChild.data
```
Since I see this here: What's the current status on the e-mail address data?

ecklbarb replied:
I think, it is not necessary, to use the id-manager here, because the id-manager was already used during the conway-codeface-analysis.

clhunsen replied:
Just leaving this as a comment, although we already talked about that: We still need the name and the e-mail address from the Codeface database to be able to align the issue data with data from other sources (e.g., mailing-list data). 
So, we need to parse the CSV file after all, read the authors and their stored e-mail addresses from the database and use those instead of the username here.

ecklbarb replied:
We have a problem with the CSV file. It doesn't include all authors. I think, they are all reporters of a issue.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+            user["username"] = assignee.attributes['username'].value
+            user["name"] = assignee.attributes['username'].value	##assignee.firstChild.data
+            user["email"] = ""
+            issue["user"] = user
```
What about `issue["author"]` (the same way as `comment["author"] = user` below)? Is there a difference here?

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+                #left = rel.getElementsByTagName('inwardlinks')
+                #right = rel.getElementsByTagName('outwardlinks')
+
+                #if (not left is None):
```
First thing: Why are these lines commented? Is the functionality not ready yet? If true, then put a TODO item here.

Second thing: Do you mean `left is not None`? I know that it is the same in Python, but not as readable. 😉

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+                relations.append(relation)
+
+            issue["relations"] = relations
+            issues.append(issue)
```
Empty line after this.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+    log.info("Syncing users with ID service...")
+
+	## auslesen aus csv-file
```
That is clearly an TODO item (written in German). 😉

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+    return issues
+
+
+def insert_user_data(issues, conf):
```
Just as a note so we do not forget: We may want to re-use the function across both issue-parsing-related files later.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+import argparse
```
In general, the indentation should be done by 4 spaces, not tabs. At least, my IDE is complaining about that.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+        sys.exit(-1)
+
+
+    person_data = list()
```
Definition is obsolete as you override the value rightaway.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+    :param issues: the issues to dump
+    :param results_folder: the folder where to place 'issues.list' output file
+    """
```
The output file's name in the doc-string is wrong. While the issues are printed in Gephi format, this information is missing in the documentation.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+    """
+
+    # construct path to output file
+    output_file = os.path.join(results_folder, "jira-issues.list")
```
Please rename the output file to "issues-jira.list". This way, the file ordering in the data folder is more consistent with the data and easier to understand.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+    # construct path to output file
+    output_file_nodes = os.path.join(results_folder, "issue_nodes.csv")
+    output_file_edges = os.path.join(results_folder, "issue_edges.csv")
```
Please rename the output files to "issues-jira-gephi-*.csv". This way, the file ordering in the data folder is more consistent with the data and easier to understand.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+
+    # write to output file
+    csv_writer.write_to_csv(output_file, lines)
+
```
Extra blank line here (PEP 8).

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+    def get_or_update_user(user, buffer_db=user_buffer):
+        # fix encoding for name and e-mail address
+        if user["name"] is not None:
+            name = unicode(user["name"]).encode("utf-8")
```
Note to self: We should use the function from `codeface.util` that I added recently for UTF-8 handling.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -0,0 +1,401 @@
+        for comment in issue["comments"]:
+            lines.append((
+                issue['externalId'],
+                issue['state'],
```
Just as a cross-reference here: I opened issue se-passau/codeface-extraction-r#91  on the statuses that can be used in Jira.

clhunsen reviewed (dismissed):
The PR looks fine basically, all the functionality is all there and works fine. I commented on many small stuff that is important in most of the cases (e.g., proper names for output files). Just add a small follow-up commit to fix the urgent issues.

Additionally, this review is a start for you to get used to our style of pull-based development. If you have any questions, do not hesitate to ask.

ecklbarb committed:
changes after pull request

renaming output files; including csv parser to get names and e-mail-adresses from the conway output file 'jira-comment-authors-with-email.csv'; add user to the codeface DB; formating source code

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>

ecklbarb committed:
replace issues with 'Github issues' in the log messages

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>

ecklbarb committed:
Set utf-8 as default encoding

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>

ecklbarb committed:
Correct comment of the function 'print_to_disk'

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>

ecklbarb committed:
Use lower case and utf-8 encoding to merge xml-file and csv-file

To merge the authors from the xml-file and the csv-file,
it is nessecary to use utf-8 encoding and lower case for
the username.

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>

ecklbarb committed:
Add original issue as two events to issues.list

Until now only the comments of a issue were writen to
the issues.list-file. From now a new issue has two events:
the opening event and writing the description of the issue
as the first comment. The opening process is adapted to process
of Github issues.

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>

clhunsen reviewed (approved):
Looks good to me, but I will give it a try in my VM first before merging.

clhunsen commented:
I am still trying to get the Codeface analysis to run in my VM, so this may take another day until merge.
